cmd/kube-spawn/setup.go

```diff
@@ -102,65 +112,230 @@ func doSetup(numNodes int, baseImage, kubeSpawnDir string) {
+		K8sRuntime: k8sruntime,
+	})
+	if outbuf == nil {
+		stderr.Fatalf("Error generating kubeadm bootstrap script")
+	}
+
+	scriptsDir := filepath.Join(kubeSpawnDir, "scripts")
+	if err := os.MkdirAll(scriptsDir, os.FileMode(0755)); err != nil {
+		stderr.Fatalf("Unable to create directory %q: %v.", kubeSpawnDir, err)
+	}
+	bootstrapScript := filepath.Join(scriptsDir, "bootstrap.sh")
+	if err := ioutil.WriteFile(bootstrapScript, outbuf.Bytes(), os.FileMode(0755)); err != nil {
+		stderr.Fatalf("Error writing script file %s: %v", bootstrapScript, err)
+	}
+}
+
+func writeKubeadmInitScript() {
+	outbuf := script.GetKubeadmInit(script.KubeadmInitOpts{
+		RuntimeRkt:   (k8sruntime == "rkt"),
+		KubeSpawnDir: kubeSpawnDir,
+	})
+	if outbuf == nil {
+		stderr.Fatalf("Error generating kubeadm init script")
+	}
+
+	scriptsDir := filepath.Join(kubeSpawnDir, "scripts")
+	if err := os.MkdirAll(scriptsDir, os.FileMode(0755)); err != nil {
+		stderr.Fatalf("Unable to create directory %q: %v.", kubeSpawnDir, err)
+	}
+	initScript := filepath.Join(scriptsDir, "init.sh")
+	if err := ioutil.WriteFile(initScript, outbuf.Bytes(), os.FileMode(0755)); err != nil {
+		stderr.Fatalf("Error writing script file %s: %v", initScript, err)
+	}
+}
+
+func writeKubeadmExtraArgs() {
+	// cgroup driver defaults to systemd on most systems, but there's
+	// an issue of runc <=1.0.0-rc2 that conflicts with --cgroup-driver=systemd,
+	// so for docker runtime, we should use legacy driver "cgroupfs".
+	switch k8sruntime {
+	case "", "docker":
+		kubeadmCgroupDriver = "cgroupfs"
+		kubeadmRuntimeEndpoint = "unix:///var/run/docker.sock"
+	case "rkt":
+		kubeadmCgroupDriver = "systemd"
+		kubeadmRuntimeEndpoint = "unix:///var/run/rktlet.sock"
+	default:
+		stderr.Fatalf("runtime %s is not supported", k8sruntime)
+	}
+
+	// K8s 1.8 or newer fails to run by default when swap is enabled.
+	// So we should disable the feature with an option "--fail-swap-on=false".
+	failSwapOn := true
+	if utils.IsK8sDev(k8srelease) {
+		failSwapOn = false
+	}
+
+	// --cgroups-per-qos should be set to false, so that we can avoid issues with
+	// different formats of cgroup paths between k8s and systemd.
+	// --enforce-node-allocatable= is also necessary.
+	outbuf := script.GetKubeadmExtraArgs(script.KubeadmExtraArgsOpts{
+		CgroupDriver:  kubeadmCgroupDriver,
+		CgroupsPerQOS: false,
+		FailSwapOn:    failSwapOn,
+	})
+	if outbuf == nil {
+		stderr.Fatalf("Error generating kubeadm init script")
+	}
+
+	outbufRuntime := script.GetKubeadmExtraRuntime(script.KubeadmExtraRuntimeOpts{
+		RktRuntime:      (k8sruntime == "rkt"),
+		RuntimeEndpoint: kubeadmRuntimeEndpoint,
+		RequestTimeout:  kubeadmRequestTimeout,
+	})
+	if outbufRuntime == nil {
+		stderr.Fatalf("Error generating kubeadm init script")
 	}
 
-	log.Printf("All nodes are running. Use machinectl to login/stop/etc.")
+	// NOTE: the 1st part "outbuf" and the 2nd part "outbufRuntime" is merged into
```

Comment:
why do we merge the unit files here? cri-o and rktlet both use `--container-runtime=remote` and the endpoint option.
Doestn't it make more sense to keep them seperate?

Reply:
@robertgzr As discussed, I also would like to keep them separate, but it was not that trivial. The thing is, 2 different definitions of `KUBELET_EXTRA_ARGS` need to be merged into 1 definition in the end. Otherwise kubelet won't be able to recognize all the options. So we decided to keep it simply in a single source file.